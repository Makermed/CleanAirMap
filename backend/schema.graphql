### This file was generated by Nexus Schema
### Do not make changes to this file directly


"""
A date-time string at UTC, such as 2007-12-03T10:15:30Z, compliant with the `date-time` format outlined in section 5.6 of the RFC 3339 profile of the ISO 8601 standard for representation of dates and times using the Gregorian calendar.
"""
scalar DateTime

"""
The `JSONObject` scalar type represents JSON objects as specified by [ECMA-404](http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf).
"""
scalar JSONObject @specifiedBy(url: "http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf")

type Location {
  geometry: JSONObject

  """The CleanAirDB ID for this feature"""
  id: Int!

  """Properties of the location"""
  properties: POIProperties!
  rooms: [Room!]

  """GeoJSON type. Always 'Feature'"""
  type: String!
}

input LocationInputType {
  geometry: JSONObject

  """Properties of the location"""
  properties: POIPropertiesInputType!

  """GeoJSON type. Always 'Feature'"""
  type: String!
}

type Mutation {
  createLocation(data: LocationInputType!): Location
  createRoom(locationId: Int!, name: String!): Room
}

type POIProperties {
  feature_type: String
  full_address: String!
  name: String!
}

input POIPropertiesInputType {
  feature_type: String
  full_address: String!
  name: String!
}

"""A GeoJSON Position"""
scalar Position

type Query {
  locations(locationId: Int): [Location]
  rooms(roomId: Int): [Room]
}

type Room {
  created_at: DateTime
  created_id: Int
  locationId: Int
  name: String
  roomId: Int
}