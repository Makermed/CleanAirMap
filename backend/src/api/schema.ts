import { makeSchema } from 'nexus'
import { join } from 'path'
import * as allTypes from './graphql'

export const schema = makeSchema({
  types: allTypes,
  outputs: {
    typegen: join(__dirname, '../..', 'nexus-typegen.ts'),
    schema: join(__dirname, '../..', 'schema.graphql'),
  },
  contextType: {
    module: join(__dirname, "./context.ts"),
    export: "Context",
  },
  sourceTypes: {
    headers: [ 'import { LocationModel } from "./src/dataAccess/dataTypes" '],
    modules: []
  }
})
