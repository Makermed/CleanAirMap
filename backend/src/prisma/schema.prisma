// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model Location {
  locationId  Int      @id @default(autoincrement())
  name        String
  type        Int
  street      String
  locality    String?
  place       String
  district    String?
  region      String
  postcode    String?
  country     String
  latitude    Float
  longitude   Float
  description String?
  avgCo2      Float?
  rooms       Room[]
  created_id  Int
  created_at  DateTime @default(now())
}

model Room {
  roomId       Int       @id @default(autoincrement())
  locationId   Int
  location     Location  @relation(fields: [locationId], references: [locationId])
  name         String
  readings     Reading[]
  created_id   Int
  created_user User      @relation(fields: [created_id], references: [id])
  created_at   DateTime  @default(now())
}

model Reading {
  id           Int      @id @default(autoincrement())
  value        Decimal
  unit         String
  notes        String?
  locationId   Int
  readingRoom  Room?    @relation(fields: [roomId], references: [roomId])
  roomId       Int?
  created_id   Int
  created_user User     @relation(fields: [created_id], references: [id])
  created_at   DateTime @default(now())
}

model User {
  id       Int       @id @default(autoincrement())
  rooms    Room[]
  readings Reading[]
}
