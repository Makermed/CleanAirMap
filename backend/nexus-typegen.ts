import { LocationModel, RoomModel } from "./src/dataAccess/dataTypes" 

import type { Context } from "./src/api/context"
import type { core } from "nexus"
declare global {
  interface NexusGenCustomInputMethods<TypeName extends string> {
    /**
     * A date-time string at UTC, such as 2007-12-03T10:15:30Z, compliant with the `date-time` format outlined in section 5.6 of the RFC 3339 profile of the ISO 8601 standard for representation of dates and times using the Gregorian calendar.
     */
    datetime<FieldName extends string>(fieldName: FieldName, opts?: core.CommonInputFieldConfig<TypeName, FieldName>): void // "DateTime";
  }
}
declare global {
  interface NexusGenCustomOutputMethods<TypeName extends string> {
    /**
     * A date-time string at UTC, such as 2007-12-03T10:15:30Z, compliant with the `date-time` format outlined in section 5.6 of the RFC 3339 profile of the ISO 8601 standard for representation of dates and times using the Gregorian calendar.
     */
    datetime<FieldName extends string>(fieldName: FieldName, ...opts: core.ScalarOutSpread<TypeName, FieldName>): void // "DateTime";
  }
}


declare global {
  interface NexusGen extends NexusGenTypes {}
}

export interface NexusGenInputs {
}

export interface NexusGenEnums {
  LocationEnum: 999 | 15 | 8 | 2 | 18 | 3 | 14 | 22 | 23 | 19 | 10 | 4 | 9 | 7 | 12 | 1 | 13 | 11 | 16 | 21 | 6 | 17 | 20 | 0 | 5 | 24
}

export interface NexusGenScalars {
  String: string
  Int: number
  Float: number
  Boolean: boolean
  ID: string
  DateTime: Date
}

export interface NexusGenObjects {
  Location: LocationModel;
  Mutation: {};
  Query: {};
  Room: RoomModel;
}

export interface NexusGenInterfaces {
}

export interface NexusGenUnions {
}

export type NexusGenRootTypes = NexusGenObjects

export type NexusGenAllTypes = NexusGenRootTypes & NexusGenScalars & NexusGenEnums

export interface NexusGenFieldTypes {
  Location: { // field return type
    avgCo2: number | null; // Float
    created_at: NexusGenScalars['DateTime'] | null; // DateTime
    created_id: number | null; // Int
    description: string | null; // String
    district: string | null; // String
    latitude: number | null; // Float
    locality: string | null; // String
    locationId: number | null; // Int
    longitude: number | null; // Float
    name: string | null; // String
    place: string | null; // String
    postcode: string | null; // String
    region: string | null; // String
    rooms: NexusGenRootTypes['Room'][] | null; // [Room!]
    street: string | null; // String
    type: NexusGenEnums['LocationEnum'] | null; // LocationEnum
  }
  Mutation: { // field return type
    createLocation: NexusGenRootTypes['Location'] | null; // Location
    createRoom: NexusGenRootTypes['Room'] | null; // Room
  }
  Query: { // field return type
    locations: Array<NexusGenRootTypes['Location'] | null> | null; // [Location]
    rooms: Array<NexusGenRootTypes['Room'] | null> | null; // [Room]
  }
  Room: { // field return type
    created_at: NexusGenScalars['DateTime'] | null; // DateTime
    created_id: number | null; // Int
    locationId: number | null; // Int
    name: string | null; // String
    roomId: number | null; // Int
  }
}

export interface NexusGenFieldTypeNames {
  Location: { // field return type name
    avgCo2: 'Float'
    created_at: 'DateTime'
    created_id: 'Int'
    description: 'String'
    district: 'String'
    latitude: 'Float'
    locality: 'String'
    locationId: 'Int'
    longitude: 'Float'
    name: 'String'
    place: 'String'
    postcode: 'String'
    region: 'String'
    rooms: 'Room'
    street: 'String'
    type: 'LocationEnum'
  }
  Mutation: { // field return type name
    createLocation: 'Location'
    createRoom: 'Room'
  }
  Query: { // field return type name
    locations: 'Location'
    rooms: 'Room'
  }
  Room: { // field return type name
    created_at: 'DateTime'
    created_id: 'Int'
    locationId: 'Int'
    name: 'String'
    roomId: 'Int'
  }
}

export interface NexusGenArgTypes {
  Mutation: {
    createLocation: { // args
      country: string; // String!
      description?: string | null; // String
      district?: string | null; // String
      lat: number; // Float!
      locality?: string | null; // String
      long: number; // Float!
      name: string; // String!
      place: string; // String!
      postcode?: string | null; // String
      region: string; // String!
      street: string; // String!
      type: NexusGenEnums['LocationEnum']; // LocationEnum!
    }
    createRoom: { // args
      locationId: number; // Int!
      name: string; // String!
    }
  }
  Query: {
    locations: { // args
      locationId?: number | null; // Int
    }
    rooms: { // args
      roomId?: number | null; // Int
    }
  }
}

export interface NexusGenAbstractTypeMembers {
}

export interface NexusGenTypeInterfaces {
}

export type NexusGenObjectNames = keyof NexusGenObjects;

export type NexusGenInputNames = never;

export type NexusGenEnumNames = keyof NexusGenEnums;

export type NexusGenInterfaceNames = never;

export type NexusGenScalarNames = keyof NexusGenScalars;

export type NexusGenUnionNames = never;

export type NexusGenObjectsUsingAbstractStrategyIsTypeOf = never;

export type NexusGenAbstractsUsingStrategyResolveType = never;

export type NexusGenFeaturesConfig = {
  abstractTypeStrategies: {
    isTypeOf: false
    resolveType: true
    __typename: false
  }
}

export interface NexusGenTypes {
  context: Context;
  inputTypes: NexusGenInputs;
  rootTypes: NexusGenRootTypes;
  inputTypeShapes: NexusGenInputs & NexusGenEnums & NexusGenScalars;
  argTypes: NexusGenArgTypes;
  fieldTypes: NexusGenFieldTypes;
  fieldTypeNames: NexusGenFieldTypeNames;
  allTypes: NexusGenAllTypes;
  typeInterfaces: NexusGenTypeInterfaces;
  objectNames: NexusGenObjectNames;
  inputNames: NexusGenInputNames;
  enumNames: NexusGenEnumNames;
  interfaceNames: NexusGenInterfaceNames;
  scalarNames: NexusGenScalarNames;
  unionNames: NexusGenUnionNames;
  allInputTypes: NexusGenTypes['inputNames'] | NexusGenTypes['enumNames'] | NexusGenTypes['scalarNames'];
  allOutputTypes: NexusGenTypes['objectNames'] | NexusGenTypes['enumNames'] | NexusGenTypes['unionNames'] | NexusGenTypes['interfaceNames'] | NexusGenTypes['scalarNames'];
  allNamedTypes: NexusGenTypes['allInputTypes'] | NexusGenTypes['allOutputTypes']
  abstractTypes: NexusGenTypes['interfaceNames'] | NexusGenTypes['unionNames'];
  abstractTypeMembers: NexusGenAbstractTypeMembers;
  objectsUsingAbstractStrategyIsTypeOf: NexusGenObjectsUsingAbstractStrategyIsTypeOf;
  abstractsUsingStrategyResolveType: NexusGenAbstractsUsingStrategyResolveType;
  features: NexusGenFeaturesConfig;
}


declare global {
  interface NexusGenPluginTypeConfig<TypeName extends string> {
  }
  interface NexusGenPluginInputTypeConfig<TypeName extends string> {
  }
  interface NexusGenPluginFieldConfig<TypeName extends string, FieldName extends string> {
  }
  interface NexusGenPluginInputFieldConfig<TypeName extends string, FieldName extends string> {
  }
  interface NexusGenPluginSchemaConfig {
  }
  interface NexusGenPluginArgConfig {
  }
}