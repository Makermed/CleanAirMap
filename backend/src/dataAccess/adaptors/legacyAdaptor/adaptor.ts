import { IDataAdaptor, ILocation, IReading, IRoom } from "../../interfaces";
import LocationDAO from "./LocationDAO";
import RoomDAO from "./RoomDAO";
import ReadingDAO from "./ReadingDAO";
import { HttpLink, ApolloClient, InMemoryCache, NormalizedCacheObject } from "@apollo/client/core";
import fetch from 'node-fetch'

/**
 * This adaptor queries from the legacy GraphQL endpoint and is just for
 * development/migration.
 */
export class LegacyAdaptor implements IDataAdaptor {
    client: ApolloClient<NormalizedCacheObject>;
    locationDAO: ILocation;
    roomDAO: IRoom;
    readingDAO: IReading;

    constructor(private endpoint: string) {
        this.client = new ApolloClient({
            link: new HttpLink({
                uri: endpoint,
                fetch: fetch
              }),
            cache: new InMemoryCache()
          });
      this.locationDAO = new LocationDAO(this.client);
      this.roomDAO = new RoomDAO(this.client);
      this.readingDAO = new ReadingDAO(this.client)
    }
  }
