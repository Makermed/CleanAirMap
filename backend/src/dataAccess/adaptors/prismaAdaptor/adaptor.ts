import { PrismaClient } from "@prisma/client";
import { IDataAdaptor, ILocation, IRoom } from "../../interfaces";
import { LocationDAO } from "./LocationDAO";
import { RoomDAO } from "./RoomDAO";

const prisma = new PrismaClient({
    log: ['query', 'info', 'warn', 'error'],
  })

export class PrismaDataAdaptor implements IDataAdaptor {
    client : PrismaClient
    locationDAO: ILocation
    roomDAO: IRoom

    constructor() {
        this.client = prisma
        this.locationDAO = new LocationDAO(this.client)
        this.roomDAO = new RoomDAO(this.client)
    }
}