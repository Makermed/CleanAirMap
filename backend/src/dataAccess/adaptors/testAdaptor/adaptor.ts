import { IDataAdaptor } from "../../IDataAdaptor";
import { ILocation } from "../../ILocation";
import { IRoom } from "../../IRoom";
import { LocationModel, RoomModel } from "../../dataTypes";

class TestLocationDAO implements ILocation {
    getById(id: number): Promise<LocationModel | null> {
        return Promise.reject(new Error("Not implemented"));
    }

    getMany(maxResults?: number | undefined): Promise<LocationModel[]> {
        return Promise.reject(new Error("Not implemented"));
    }

    getRooms(locationId: number): Promise<RoomModel[] | null> {
        return Promise.reject(new Error("Not implemented"));
    }

    create(reading: LocationModel): Promise<LocationModel | null> {
        return Promise.reject(new Error("Not implemented"));
    }
}

class TestRoomDAO implements IRoom {
    create(reading: RoomModel): Promise<RoomModel | null> {
        return Promise.reject(new Error("Not implemented"));
    }
    getById(id: number): Promise<RoomModel | null> {
        return Promise.reject(new Error("Not implemented"));
    }
    getByLocation(location: LocationModel): Promise<RoomModel[] | null> {
        return Promise.reject(new Error("Not implemented"));
    }
    getMany(maxResults?: number | undefined): Promise<RoomModel[]> {
        return Promise.reject(new Error("Not implemented"));
    }
}

export class TestAdaptor implements IDataAdaptor {
    locationDAO: ILocation;
    roomDAO: IRoom;

    constructor() {
        this.locationDAO = new TestLocationDAO();
        this.roomDAO = new TestRoomDAO();
    }
}