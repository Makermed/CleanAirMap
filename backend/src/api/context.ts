import { IDataAdaptor } from "../dataAccess/interfaces";
import { PrismaDataAdaptor } from "../dataAccess/adaptors/prismaAdaptor/adaptor";

export interface Context {
    db: IDataAdaptor
}

export const createContext =async () => ({
    db: new PrismaDataAdaptor()
})