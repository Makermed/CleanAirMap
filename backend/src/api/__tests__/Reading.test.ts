import { ApolloServer } from "@apollo/server";
import { schema } from "../schema";
import { Context } from "../context"
import { TestAdaptor } from "../../dataAccess/adaptors/testAdaptor/adaptor";
import { NexusGenInputs } from "../../../nexus-typegen";
import * as testSetup from './__test_setup'
import { queryForSingleResult } from './__query_helper';

const testServer = new ApolloServer<Context>({schema})

const testContext = {
    db: new TestAdaptor()
}

describe('add reading and location mutation', () => {
  testSetup.mockLocationDAO(testContext.db);
  testSetup.mockReadingDAO(testContext.db);
  it('stores a new location with a new reading' , async () => {
    const inputData : NexusGenInputs["AddReadingNewLocationInputType"] = {
      location : {
        type: 'Feature',
        geometry: { coordinates: [-86.9259644,43.4708553] },
        properties: {
          name: 'Create location test',
          feature_type: 'poi',
          full_address: '567 Kings Boul., SomePlace, Ontario, M4G 2E5, Canada'
        }
      },
      reading : {
        value: 567,
        //@ts-ignore Type conflict on unit Enum, not sure why
        unit: "PPM_CO2",
        notes: "Reading notes"
      }
    };

    const result = await queryForSingleResult(testServer, CREATE_LOCATION_WITH_READING_QUERY, { data: inputData }, testContext);

    expect(result.errors).toBeUndefined();
    expect(result.data).not.toBeNull();
    expect(result.data?.addReadingNewLocation).toMatchSnapshot();
  })
})

const READINGS_QUERY =
`query Locations($locationId: Int) {
  locations(locationId: $locationId) {
    type
    id
    geometry
    properties {
      name
      full_address
      feature_type
    }
  }
}`

const CREATE_LOCATION_WITH_READING_QUERY =
`mutation CreateLocationWithReading($data: AddReadingNewLocationInputType!) {
  addReadingNewLocation(input: $data) {
    location {
      type
      id
      geometry
      properties {
        feature_type
        full_address
        name
      }
    }
    reading {
      id
      location {
        id
      }
      value
      unit
      notes
      created_at
    }
  }
}`