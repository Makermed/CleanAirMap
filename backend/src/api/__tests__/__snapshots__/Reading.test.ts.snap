// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add reading and location mutation stores a new location with a new reading 1`] = `
{
  "location": {
    "geometry": {
      "coordinates": [
        -86.9259644,
        43.4708553,
      ],
    },
    "id": 1,
    "properties": {
      "feature_type": "poi",
      "full_address": "567 Kings Boul., SomePlace, Ontario, M4G 2E5, Canada",
      "name": "Create location test",
    },
    "type": "Feature",
  },
  "reading": {
    "created_at": 2023-01-05T00:00:00.000Z,
    "id": 1,
    "location": {
      "id": 1,
    },
    "notes": "Reading notes",
    "unit": "PPM_CO2",
    "value": 567,
  },
}
`;